overlord/managers_test.go

```diff
@@ -5804,6 +5819,265 @@ func (s *mgrsSuite) TestRemodelReregistration(c *C) {
+				"default-channel": "20",
+			},
+			map[string]interface{}{
+				"name":     "foo",
+				"presence": "required",
+			},
+			map[string]interface{}{
+				"name":     "bar",
+				"presence": "required",
+			}},
+		"revision": "1",
+	})
+
+	// mock the modeenv file
+	m := boot.Modeenv{
+		Mode:                      "run",
+		Base:                      "core20_3.snap",
+		CurrentKernels:            []string{"pc-kernel_2.snap"},
+		CurrentRecoverySystems:    []string{"1234"},
+		GoodRecoverySystems:       []string{"1234"},
+		CurrentKernelCommandLines: []string{"snapd_recovery_mode=run mocked static"},
+	}
+	err = m.WriteTo("")
+	c.Assert(err, IsNil)
+	c.Assert(s.o.DeviceManager().ReloadModeenv(), IsNil)
+
+	err = bl.SetBootVars(map[string]string{
+		"snap_kernel": "pc-kernel_2.snap",
+	})
+	c.Assert(err, IsNil)
+
+	chg, err := devicestate.Remodel(st, newModel)
+	c.Assert(err, IsNil)
+
+	c.Check(devicestate.Remodeling(st), Equals, true)
+
+	st.Unlock()
+	err = s.o.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil, Commentf(s.logbuf.String()))
+
+	c.Check(chg.Status(), Equals, state.DoingStatus, Commentf("remodel change failed: %v", chg.Err()))
+	c.Check(devicestate.Remodeling(st), Equals, true)
+	restarting, kind := st.Restarting()
+	c.Check(restarting, Equals, true)
+	c.Assert(kind, Equals, state.RestartSystemNow)
+
+	expectedLabel := time.Now().Format("20060102")
+
+	vars, err := bl.GetBootVars("try_recovery_system", "recovery_system_status")
+	c.Assert(err, IsNil)
+	c.Assert(vars, DeepEquals, map[string]string{
+		"try_recovery_system":    expectedLabel,
+		"recovery_system_status": "try",
+	})
+
+	// the new required-snap "foo" is not installed yet
+	var snapst snapstate.SnapState
+	err = snapstate.Get(st, "foo", &snapst)
+	c.Assert(err, Equals, state.ErrNoState)
+
+	// simulate successful reboot to recovery and back
+	state.MockRestarting(st, state.RestartUnset)
+	// this would be done by snap-bootstrap in initramfs
+	err = bl.SetBootVars(map[string]string{
+		"try_recovery_system":    expectedLabel,
+		"recovery_system_status": "tried",
+	})
+	c.Assert(err, IsNil)
+
+	// reset, so that after-reboot handling of tried system is executed
+	s.o.DeviceManager().ResetTriedRecoverySystemsRan()
+	st.Unlock()
+	err = s.o.DeviceManager().Ensure()
+	st.Lock()
+	c.Assert(err, IsNil)
+
+	st.Unlock()
+	err = s.o.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil)
+
+	c.Assert(chg.Status(), Equals, state.DoneStatus, Commentf("remodel change failed: %v", chg.Err()))
+	// boot variables are cleared
+	vars, err = bl.GetBootVars("try_recovery_system", "recovery_system_status")
+	c.Assert(err, IsNil)
+	c.Assert(vars, DeepEquals, map[string]string{
+		"try_recovery_system":    "",
+		"recovery_system_status": "",
+	})
+
+	for _, name := range []string{"core20", "pc-kernel", "pc", "snapd", "foo", "bar"} {
+		var snapst snapstate.SnapState
+		err = snapstate.Get(st, name, &snapst)
+		c.Assert(err, IsNil)
+	}
+
+	// ensure sorting is correct
```

Comment:
With the exception that verifying tasks before  we run it will be slightly misleading as the snap tasks will not have materialized yet as ensure prepreq did not run yet.

Reply:
so you left it as is, that's fine, thanks for the explanation